google/cloud/spanner/value.cc

```diff
@@ -88,6 +88,79 @@ bool Equal(google::spanner::v1::Type const& pt1,
   }
 }
 
+constexpr char kArrayFieldDelimiter[] = ", ";    // NOLINT
+constexpr char kStructFieldDelimiter[] = " | ";  // NOLINT
+
+std::ostream& StreamHelper(std::ostream& os, google::protobuf::Value const& v,
+                           google::spanner::v1::Type const& t);
+
+std::ostream& StructArrayStreamHelper(
+    std::ostream& os, google::protobuf::Value const& v,
+    std::function<std::tuple<char, const char*, char,
+                             google::spanner::v1::Type>(int)> const&
+        delimiters_type_function) {
+  int i = 0;
+  os << std::get<0>(delimiters_type_function(i));
+  if (v.list_value().values_size() > 0) {
+    for (; i < v.list_value().values_size() - 1; ++i) {
+      StreamHelper(os, v.list_value().values(i),
+                   std::get<3>(delimiters_type_function(i)))
+          << std::get<1>(delimiters_type_function(i));
+    }
+    return StreamHelper(os, v.list_value().values(i),
+                        std::get<3>(delimiters_type_function(i)))
+           << std::get<2>(delimiters_type_function(i));
+  }
+  return os << std::get<2>(delimiters_type_function(i));
+}
+
+std::ostream& StreamHelper(std::ostream& os, google::protobuf::Value const& v,
+                           google::spanner::v1::Type const& t) {
+  if (v.kind_case() == google::protobuf::Value::kNullValue &&
+      t.code() != google::spanner::v1::ARRAY &&
+      t.code() != google::spanner::v1::STRUCT) {
+    return os;
```

Comment:
Q. Should this render "NULL"?

Reply:
Perhaps, but for a string field how would I differentiate between a null value and a literal of "NULL"?

Reply:
Perhaps a string field renders with '"' delimiters (just like structs and arrays have delimiters)?